### PR TITLE
fix(sdk-core): add EVM staking types to TSS signing bypass list

### PR DESCRIPTION
**Pull request by mrdanish26**
TICKET: WAL-756

## Summary

  **1. WAL-756 — BNB/BSC staking broken**
  Staking wallets call `signTransaction` without `txParams`. After WAL-375, `resolveEffectiveTxParams()` throws `InvalidTransactionError` because `effectiveTxParams.type` is undefined and staking intentType strings were not in `NO_RECIPIENT_TX_TYPES`.

  **2. XDC/BSC token enable/approve**
Coin-specific `verifyTssTransaction` overrides in `xdcToken.ts`, `bscToken.ts`, `xdc.ts`, `bsc.ts`, and `evmCoin.ts` had stale hardcoded bypass lists and never updated. When the parent class added `tokenApproval`, `enableToken`, `bridgeFunds`, etc., these overrides were left behind — causing `"missing txParams"` errors on app.bitgo-test.com when enabling tokens on XDC.

  ## Solution

  **Staking fix (`recipientUtils.ts`):**
  - Detect staking intents generically via `stakingRequestId` presence on
    the intent (a required field on `BaseStakeIntent` in `@bitgo/public-types`)
    rather than enumerating staking intentType strings that drift over time
  - Add `intent.intentType` fallback so `txType` propagates to downstream
    `verifyTssTransaction` callers when `txParams.type` is absent

  **Centralization (all coin modules):**
  - Export `NO_RECIPIENT_TX_TYPES` from sdk-core public API (`tss/index.ts`)
    as the single source of truth
  - Replace all stale inline bypass arrays with `NO_RECIPIENT_TX_TYPES.has()`
    in: `abstractEthLikeNewCoins.ts`, `bsc.ts`, `bscToken.ts`, `xdc.ts`,
    `xdcToken.ts`, `evmCoin.ts`
  - Existing `verifyTssTransaction` overrides preserved

  ## Testing

  - [ ] Unit tests updated 
  - [ ] `staking-bnb-hot` E2E test in bitgo-retail passes after SDK bump

### CODE REVIEW

File: modules/sdk-coin-bsc/src/bsc.ts [hunk 1]
```diff
@@ -54,34 +50,4 @@ export class Bsc extends AbstractEthLikeNewCoins {
     const explorerUrl = common.Environments[this.bitgo.getEnv()].bscscanBaseUrl;
     return await recoveryBlockchainExplorerQuery(query, explorerUrl as string, apiToken);
   }
-
```
Comment (zahin-mohammad) [inline modules/sdk-coin-bsc/src/bsc.ts:57]: Context for this removal?

File: modules/abstract-eth/src/abstractEthLikeNewCoins.ts [hunk 2]
```diff
@@ -3118,6 +3118,14 @@ export abstract class AbstractEthLikeNewCoins extends AbstractEthLikeCoin {
...
             'enableToken',
             'customTx',
+            // EVM staking (BSC/BNB, ETH, CELO) — mirrors NO_RECIPIENT_TX_TYPES in recipientUtils.ts
+            'stakingLock',
+            'stakingVote',
+            'stakingActivate',
+            'stakingUnvote',
+            'stakingUnlock',
+            'stakingWithdraw',
+            'stakingDeactivate',
```
Comment (zahin-mohammad) [inline modules/abstract-eth/src/abstractEthLikeNewCoins.ts]: Not great that we are duplicating this same list in multiple areas/modules. Can we de-duplicate as a follow up?

File: modules/sdk-coin-bsc/src/bsc.ts (hunk 1 above)
Comment (mrdanish26) [reply modules/sdk-coin-bsc/src/bsc.ts:57]: The original bsc.ts had its own `verifyTssTransaction` override that duplicated the parent's logic but only knew about 3 types:
// OLD bsc.ts — only 3/15 types: ['acceleration', 'fillNonce', 'transferToken'].includes(txParams.type)

This override shadowed the parent (AbstractEthLikeNewCoins), so even though the parent now has the correct 15-type list, BSC never saw it,  the subclass's override ran instead.

Removing it means BSC falls through to the parent, which has the complete correct list. It's the same pattern 21 of the other 23 EVM coins already use — no override at all.

File: modules/abstract-eth/src/abstractEthLikeNewCoins.ts (hunk 2 above)
Comment (mrdanish26) [reply modules/abstract-eth/src/abstractEthLikeNewCoins.ts]: Will fix this in a follow up

**Review by zahin-mohammad — DISMISSED** [review]

**Comment by lokesh-bitgo** [comment]
@claude Please do a thorough review of this PR.

Focus especially on:
- any bugs or incorrect logic in the changes
- whether these changes could impact existing flows or cause regressions
- whether the implementation is aligned with the requirement/design
- whether the changes are correct and complete
- missing edge cases, validation gaps, and test coverage gaps

Please call out concrete issues clearly and separate them by severity if possible.

**Review by ashutoshkumar-6 — COMMENTED** [review]
Overall the direction of centralizing `NO_RECIPIENT_TX_TYPES` and the staking detection via `stakingRequestId` is good. However, there is a critical gap: the fix in `resolveEffectiveTxParams` prevents `InvalidTransactionError` from throwing, but the downstream `verifyTssTransaction` guards in **every affected coin** (`bsc.ts`, `bscToken.ts`, `xdc.ts`, `xdcToken.ts`, `evmCoin.ts`, and the parent `abstractEthLikeNewCoins.ts`) will still throw `'missing txParams'` for staking types — see inline comments for details.

File: modules/sdk-core/src/bitgo/utils/tss/recipientUtils.ts [hunk 3]
```diff
@@ -43,7 +49,21 @@ export function resolveEffectiveTxParams(
...
+  if (!effectiveTxParams.type && txType) {
+    effectiveTxParams.type = txType;
+  }
+
+  // All staking intents (BSC delegate/undelegate, CELO stake/unstake, etc.) carry
+  // stakingRequestId as a required field on BaseStakeIntent (@bitgo/public-types).
+  // Use its presence as a generic staking signal — no need to enumerate every intentType.
+  const isStakingIntent = !!(txRequest.intent as any)?.stakingRequestId;
+
+  if (!effectiveTxParams.recipients?.length && !isStakingIntent && !NO_RECIPIENT_TX_TYPES.has(txType)) {
```
Comment (ashutoshkumar-6) [inline modules/sdk-core/src/bitgo/utils/tss/recipientUtils.ts:66]: **[Critical]** This fix prevents `InvalidTransactionError` here, but the staking bypass does **not** propagate to `verifyTssTransaction`.

The full call chain after this function returns for a staking intent (`stakingRequestId` present, no recipients) is:
```
resolveEffectiveTxParams returns { type: 'delegate' }  ← fixed here
  → baseCoin.verifyTransaction({ txParams: { type: 'delegate' } })
  → verifyTssTransaction: !txParams?.recipients && !NO_RECIPIENT_TX_TYPES.has('delegate')
  → throws 'missing txParams'  ← NOT fixed
```

`'delegate'`, `'undelegate'`, `'stake'`, `'unstake'` are **not** in `NO_RECIPIENT_TX_TYPES`, so every `verifyTssTransaction` override (`bsc.ts:79`, `bscToken.ts:57`, `xdc.ts:79`, `xdcToken.ts:81`, `evmCoin.ts:119`, `abstractEthLikeNewCoins.ts:3112`) will still throw for BSC staking.

The fix moves the error from `InvalidTransactionError` (here) to `Error('missing txParams')` (in `verifyTssTransaction`) but does not eliminate it. The staking flow remains broken end-to-end.

File: modules/sdk-coin-bsc/src/bsc.ts [hunk 4]
```diff
@@ -68,10 +76,7 @@ export class Bsc extends AbstractEthLikeNewCoins {
     const { txParams, txPrebuild, wallet } = params;
     if (
       !txParams?.recipients &&
-      !(
-        txParams.prebuildTx?.consolidateId ||
-        (txParams.type && ['acceleration', 'fillNonce', 'transferToken', 'tokenApproval'].includes(txParams.type))
-      )
+      !(txParams.prebuildTx?.consolidateId || (txParams.type && NO_RECIPIENT_TX_TYPES.has(txParams.type)))
```
Comment (ashutoshkumar-6) [inline modules/sdk-coin-bsc/src/bsc.ts:79]: **[Critical]** This guard still throws for BSC staking transactions.

After `resolveEffectiveTxParams` returns `effectiveTxParams = { type: 'delegate' }` (no recipients), this evaluates as:
- `!txParams?.recipients` → `true`
- `txParams.prebuildTx?.consolidateId` → `undefined` (no prebuildTx on staking tx)
- `NO_RECIPIENT_TX_TYPES.has('delegate')` → `false`
- Result: **throws `'missing txParams'`**

The staking types propagated via `intent.intentType` are not in `NO_RECIPIENT_TX_TYPES`. Same issue exists in `bscToken.ts`, `xdc.ts`, `xdcToken.ts`, and `evmCoin.ts:119`.

Possible fixes:
1. Check for a staking signal here too (e.g. set a flag on `effectiveTxParams` like `effectiveTxParams.isStakingIntent = true` in `resolveEffectiveTxParams`, then check it here), or
2. Pass staking types through `NO_RECIPIENT_TX_TYPES` with a comment, or
3. Remove this override entirely and let the parent handle it (as the commit message `467ff30d` says was intended — but the override is still present in the PR)

File: modules/abstract-eth/src/abstractEthLikeNewCoins.ts [hunk 5]
```diff
@@ -3108,17 +3109,7 @@ export abstract class AbstractEthLikeNewCoins extends AbstractEthLikeCoin {
...
-            'acceleration',
-            'fillNonce',
-            'transferToken',
-            'tokenApproval',
-            'consolidate',
-            'bridgeFunds',
-            'enableToken',
-            'customTx',
-          ].includes(txParams.type))
+        (txParams.type && NO_RECIPIENT_TX_TYPES.has(txParams.type))
```
Comment (ashutoshkumar-6) [inline modules/abstract-eth/src/abstractEthLikeNewCoins.ts:3112]: **[Critical]** Same issue as in `bsc.ts`: `NO_RECIPIENT_TX_TYPES` does not contain staking types (`delegate`, `undelegate`, `stake`, `unstake`, etc.). When a staking transaction arrives here with `txParams = { type: 'delegate' }` (returned by `resolveEffectiveTxParams`), this check throws `'missing txParams'`.

Note that this parent class is also what `bsc.ts` would fall through to if the override were removed (as the commit message for `467ff30d` states). Either way, the staking bypass needs to be present at this layer too.

File: modules/sdk-core/src/bitgo/utils/tss/recipientUtils.ts [hunk 6]
```diff
@@ -43,7 +49,21 @@ export function resolveEffectiveTxParams(
...
+
+  // Propagate the resolved type so downstream callers (e.g. verifyTssTransaction) can use it.
+  if (!effectiveTxParams.type && txType) {
+    effectiveTxParams.type = txType;
+  }
+
+  // All staking intents (BSC delegate/undelegate, CELO stake/unstake, etc.) carry
+  // stakingRequestId as a required field on BaseStakeIntent (@bitgo/public-types).
+  // Use its presence as a generic staking signal — no need to enumerate every intentType.
+  const isStakingIntent = !!(txRequest.intent as any)?.stakingRequestId;
```
Comment (ashutoshkumar-6) [inline modules/sdk-core/src/bitgo/utils/tss/recipientUtils.ts:64]: **[High]** `as any` cast is unsafe. `PopulatedIntent` in `baseTypes.ts` does not define `stakingRequestId`, so TypeScript cannot verify this field exists at compile time. If `BaseStakeIntent` in `@bitgo/public-types` renames or removes `stakingRequestId`, this silently returns `false` at runtime with no compile-time error.

Consider importing `BaseStakeIntent` from `@bitgo/public-types` and writing a proper type guard:
```typescript
function isBaseStakeIntent(intent: unknown): intent is BaseStakeIntent {
  return typeof (intent as any)?.stakingRequestId === 'string';
}
const isStakingIntent = isBaseStakeIntent(txRequest.intent);
```
Or extend the `PopulatedIntent` union in `baseTypes.ts` to include a staking variant.

File: modules/abstract-eth/src/abstractEthLikeNewCoins.ts (hunk 5 above)
Comment (mohammadalfaiyazbitgo) [reply modules/abstract-eth/src/abstractEthLikeNewCoins.ts:3112]: @mrdanish26 let's also add the `stakingRequestId`, it should be in the `prebuildTx` object. @noel-bitgo if you can confirm.

**Review by mohammadalfaiyazbitgo — COMMENTED** [review]
please address conflicts

**Review by lokesh-bitgo — COMMENTED** [review]
> please address conflicts

+1
